### DISCUSSION TITLE
Reusing SearchElement

### DISCUSSION THREAD

**Discussion by lufton** [getgauge/taiko#2586] [General]
I have a step definition that is failing:
```
step("Enter <value> into <fieldLabel> text field", async function(value, fieldLabel) {
	let txtBox = await textBox(below(text(fieldLabel)), near(text(fieldLabel)));
	await scrollTo(txtBox, { blockAlignment: 'center', inlineAligment: 'center' });
	await write(value, into(txtBox));
});
```

But it fails at `await write(value, into(txtBox))` line:
> Error: TextBox below Element with text THE TEXT NEXT TO ELEMENT  and near Element with text THE TEXT NEXT TO ELEMENT  not found

This is weird, because `await scrollTo(txtBox, { blockAlignment: 'center', inlineAligment: 'center' })` works as expected, so there is corresponding element.
Also if I replace `txtBox` with `textBox(below(text(fieldLabel)), near(text(fieldLabel)))` inside `scrollTo` and `write` method calls, then this code words. So it looks like as soon as I used txtBox variable it is not holding actual element anymore.

**Comment by NivedhaSenthil** [comment]
Hi, Can you please give us a sample to replicate this ?

**Reply by lufton** [reply]
Yes, @NivedhaSenthil, I think it is only reproducible when using RelativeSelector:
```
openBrowser();
goto("https://gauge.org/gauge-taiko/");
let lnk = await link("Taiko", below("Explore")); // Footer menu item
await scrollTo(lnk);
await click(lnk);
```

**Reply by NivedhaSenthil** [reply]
huh thanks for the snippet, I can sense what might be happening here. My guess is since the relative selector is calculated before scrolling the position is changed the element becomes invalid in the later action. I think we should be updating the relative selector to recalculate the bounding box if there is no match found with the existing one. Creating an issue from this to track on it. Thanks for reporting this.

**Reply by lufton** [reply]
Done #2590

**Comment by zabil** [comment]
> My guess is since the relative selector is calculated before scrolling the position is changed the element becomes invalid in the later action.

Good catch. I think this is a risk with all selectors unless there is some sort of mechanism to lazily evaluate the positions whenever it is used.

While not ideal, the script should be not assign selectors to variables as the page can change state for other actions too. For example.

```
openBrowser();
goto("https://gauge.org/gauge-taiko/");
let lnk = await link("Taiko", below("Explore")); // Footer menu item
await goto('url');
await click(lnk);
```
